include/revng/Model/TupleTree.h

```diff
@@ -1043,22 +1056,32 @@ class TupleTreeReference {
 
   const TupleTreePath &path() const { return Path; }
 
-  T *get() {
-    revng_check(Root != nullptr);
-
-    if (Path.size() == 0)
-      return nullptr;
-
-    return getByPath<T>(Path, *Root);
+  bool hasNullRoot() const {
+    const auto IsNullVisitor = [](const auto &Ptr) { return Ptr == nullptr; };
+    return std::visit(IsNullVisitor, Root);
   }
 
   const T *get() const {
-    revng_check(Root != nullptr);
+    revng_check(Root.index() != std::variant_npos);
+    revng_check(not hasNullRoot());
 
     if (Path.size() == 0)
       return nullptr;
 
-    return getByPath<T>(Path, *Root);
+    const auto GetByPathVisitor =
+      [&Path = std::as_const(Path)](const auto &RootPointer) {
+        return getByPath<T>(Path, *RootPointer);
+      };
+
+    return std::visit(GetByPathVisitor, Root);
+  }
+
+  T *get() {
+    // Strict check that we're not holding the const alternative
+    revng_check(not std::holds_alternative<const RootT *>(Root));
+    // Then just call the const version and cast constness away.
+    auto *const_this = const_cast<const TupleTreeReference<T, RootT> *>(this);
+    return const_cast<T *>(const_this->get());
```

Comment:
Not really. If anything, we could do:
```C++
return getByPath<T>(Path, std::get<RootT *>(Root));
```

But in this way we wouldn't be sharing the implementation with the other `get()` method, which I thought was kinda nice because it guarantees they do the same thing.

But yes, the `const_cast` dance is not really nice to see. I can go with the other version up here if you prefer it.

Reply:
In the meantime, to explore our options, I tried to go also with this:
```c++
    T *get() {
      revng_check(Root.index() != std::variant_npos);
      revng_check(not hasNullRoot());

      if (Path.size() == 0)
        return nullptr;

      return getByPath<T>(Path, std::get<RootT *>(Root));
    }
```
and I get assertions from `DwarfImporter` for trying to use `llvm::isa` on a null pointer.

This is slightly puzzling, I'm investigating it.

Reply:
Ok it was just a bug of mine. I'll leave the `const_cast` dance out then, and duplicate a little bit of code.

Reply:
Done, take a look!